pkg/controller/extensions/extension/extension.go

```diff
@@ -0,0 +1,251 @@
+	return reconcile.Result{}, nil
+}
+
+func (h *extensionHandler) processRBAC(ctx context.Context) error {
+	if len(h.i.Spec.Permissions.Rules) == 0 {
+		return nil
+	}
+
+	// create service account
+	sa := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            h.i.Name,
+			Namespace:       h.i.Namespace,
+			OwnerReferences: []metav1.OwnerReference{h.i.OwnerReference()},
+		},
+	}
+	if err := h.kube.Create(ctx, sa); err != nil && !kerrors.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create service account: %+v", err)
+	}
+
+	// create role
+	cr := &rbac.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            h.i.Name,
+			OwnerReferences: []metav1.OwnerReference{h.i.OwnerReference()},
+		},
+		Rules: h.i.Spec.Permissions.Rules,
+	}
+	if err := h.kube.Create(ctx, cr); err != nil && !kerrors.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create cluster role: %+v", err)
+	}
+
+	// create rolebinding between service account and role
+	crb := &rbac.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            h.i.Name,
+			OwnerReferences: []metav1.OwnerReference{h.i.OwnerReference()},
+		},
+		RoleRef: rbac.RoleRef{APIGroup: rbac.GroupName, Kind: "ClusterRole", Name: h.i.Name},
+		Subjects: []rbac.Subject{
+			{Name: h.i.Name, Namespace: h.i.Namespace, Kind: rbac.ServiceAccountKind},
+		},
+	}
+	if err := h.kube.Create(ctx, crb); err != nil && !kerrors.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create cluster role binding: %+v", err)
+	}
+
+	return nil
+}
+
+func (h *extensionHandler) processDeployment(ctx context.Context) error {
+	controllerDeployment := h.i.Spec.Controller.Deployment
+	if controllerDeployment == nil {
+		return nil
+	}
+
+	// ensure the deployment is set to use this extension's service account that we created
+	deploymentSpec := *controllerDeployment.Spec.DeepCopy()
+	deploymentSpec.Template.Spec.ServiceAccountName = h.i.Name
+
+	d := &apps.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            controllerDeployment.Name,
+			Namespace:       h.i.Namespace,
+			OwnerReferences: []metav1.OwnerReference{h.i.OwnerReference()},
+		},
+		Spec: deploymentSpec,
+	}
+	if err := h.kube.Create(ctx, d); err != nil && !kerrors.IsAlreadyExists(err) {
+		return fmt.Errorf("failed to create deployment: %+v", err)
+	}
+
+	// save a reference to the extension's controller
+	h.i.Status.ControllerRef = &corev1.ObjectReference{
+		APIVersion: d.APIVersion,
+		Kind:       d.Kind,
+		Name:       d.Name,
+		Namespace:  d.Namespace,
+		UID:        d.ObjectMeta.UID,
+	}
+
+	return nil
+}
+
+// fail - helper function to set fail condition with reason and message
+func fail(ctx context.Context, kube client.StatusClient, i *v1alpha1.Extension, reason, msg string) (reconcile.Result, error) {
+	log.V(logging.Debug).Info("failed extension", "i", i.Name, "reason", reason, "message", msg)
+	i.Status.SetFailed(reason, msg)
+	i.Status.UnsetCondition(corev1alpha1.Ready)
```

Comment:
What are your thoughts on "update" use-case, when the application is in a READY state, but the update has failed, hence FAILED condition is activated, but the application is still up and running (in the pre-update state), i.e. READY? Should we unset READY state?

Reply:
good question, i think we'll have to think through the specific failure scenario and it could be a case by case basis - i.e. some failures during update may indeed put the object no longer in a ready state so clearing ready is correct, but other failures might leave the object in a ready state.